objc/notify.m

```diff
@@ -13,145 +13,139 @@
 // invariant: this function should be called at most once and before `sendNotification`
 BOOL setApplication(NSString* newbundleIdentifier)
 {
-    @autoreleasepool
+    if (!installNSBundleHook())
     {
-        if (!installNSBundleHook())
-        {
-            return NO;
-        }
-        if (LSCopyApplicationURLsForBundleIdentifier((CFStringRef)newbundleIdentifier, NULL) != NULL)
-        {
-            [fakeBundleIdentifier release]; // Release old value - nil is ok
-            fakeBundleIdentifier = newbundleIdentifier;
-            [newbundleIdentifier retain]; // Retain new value - it outlives this scope
-
-            return YES;
-        }
         return NO;
     }
+
+    if (LSCopyApplicationURLsForBundleIdentifier((__bridge CFStringRef)newbundleIdentifier, NULL) != NULL)
```

Comment:
I admit I had to read up on `__bridge` a little, but I'm not entirely sure I understood this completely yet.
is the `NSString` that comes in via the argument already being reference counted? Does this really automatically free the `fakeBundleIdentifier` only when overwriting or on program-exit?
Just want to understand a bit better, as I never worked with ARC in objective-c before. Thanks

Reply:
>  is the NSString that comes in via the argument already being reference counted?

Yep. Most Objective-C objects are reference counted, as in this case the `objc_foundation::NSString` type (done via `Id<T, Owned>`). On the Rust side when that drops it will call `objc_release`.

Worth pointing out that ARC here is automatic reference counting, not atomic (like Rust's `Arc`). So objects are always reference counted with or without the feature, its just a matter of the compiler inserting the balancing where needed instead of the programmer when the feature is enabled.

 > Does this really automatically free the fakeBundleIdentifier only when overwriting or on program-exit?

I think the only time it will be freed is when overwriting it with a new value[^1] and releasing the old value. From the perspective of the ObjectiveC compiler though, `fakeBundleIdentifier` can never have its place freed because its a global and can be accessed as long as the runtime is alive (comes from the `__bundleIdentifier` swizzle). It will die on program exit, with an obligated link to [the ultimate in ~~garbage collection~~ automatic freeing](https://devblogs.microsoft.com/oldnewthing/20180228-00/?p=98125).

[^1]: This can't actually happen when the whole library is used together though because `setApplication` is guarded by a `Once`.